### PR TITLE
Multinomial sampling fix on mps for non contiguous tensors

### LINKED ISSUE #141457: distributions.categorical() produces incorrect results on mps device
### 🐛 Describe the bug

I have a neural network that outputs a tensor of size (B,4,100), with the aim of producing Bx4 categorial distributions of length 100. After some experimentation, I have figured out that `torch.distributions.Categorical` produces different output depending on if this distribution is sampled on my Macbook's `mps` device or via the CPU. The `mps` device seems to always produce identical outputs when it is sampled.

I have created a [very small repository](https://github.com/airza/torch_bug) that reproduces the issue. Running `probs.py` imports and creates a distribution based on a tensor that is either running on the `mps` or `cpu` backend. For some reason, the `mps` background always produces an identical output for each run, while the `cpu` backend appears to sample correctly. The following is an example output from the `probs.py` script, which i hope is relatively self-explanatory:

entropy
tensor([[4.5363, 4.5379, 4.5646, 4.5854]], device='mps:0')
tensor([[4.5363, 4.5378, 4.5646, 4.5854]])
mpu samples
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
cpu samples
tensor([[16, 99, 90, 16]])
tensor([[79, 94, 30, 82]])
tensor([[81, 94, 27, 49]])
tensor([[93, 13, 80,  2]])
tensor([[68, 41, 40, 37]])
tensor([[85, 67, 80, 27]])
tensor([[54, 35,  0, 65]])
tensor([[81, 42,  7, 88]])
tensor([[24, 61, 50, 11]])
tensor([[73, 67, 83, 72]])

I am not sure what is causing the problem. Am I correct in assuming this is a bug in the MPS backend rather than in my code somehow?

PS: believe my understanding of sampling a multidimensional `Categorical` distribution is correct: invoking `dist()` on a tensor of size M x N x P will normalize each row so that the sum along P is 1 and then sample independently to produce a tensor of size M x N. I would like to update the documentation so that this is more clear to others, but I wanted to verify this is correct first.

### Versions

PyTorch version: 2.5.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: version 3.27.3
Libc version: N/A

Python version: 3.12.0 (v3.12.0:0fb18b02c8, Oct  2 2023, 09:45:56) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] numpy==2.1.3
[pip3] torch==2.5.1
[conda] numpy                     1.24.2          py311h60f8152_0    conda-forge
[conda] pytorch                   2.3.1                  py3.11_0    pytorch
[conda] torchvision               0.18.1                py311_cpu    pytorch

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by Isalia20**
Fixes #141457 

As for the tests. I looked in `test/test_mps.py` but I saw that `test_multinomial` function is disabled. Glad to add test where needed if there is some place where multinomial function is tested on metal.

### CODE REVIEW

**Comment by Isalia20** [comment]
Not sure why some workflows are failing, shouldn't be related to this change I think

**Comment by Isalia20** [comment]
anyone up for review?

**Comment by Isalia20** [comment]
bump

**Comment by Isalia20** [comment]
Upon further inspection I found out that the bug was related to allocating q tensor with empty_like which lead to allocating it without contiguous block, which lead to inplace exponential not working. This solution is much cleaner now. @bdhirsh  do you know anyone who can review this?

**Comment by malfet** [comment]
@Isalia20 can you please add a test? Otherwise, looks good to me

**Comment by Isalia20** [comment]
Added the test for checking it

**Review by malfet — APPROVED** [review]
Thank you for the fix, though it looks like the test added will pass before and after the change you are proposing:
```
>>> probs = torch.tensor([[.25,.1],[.25,.1],[.25,.1],[.25,.7]]).T.to("mps")
>>> samples = {torch.multinomial(probs, 1).flatten()[0].item() for _ in range(200)}
>>> samples
{0, 1, 2, 3}
```

Oh, I see, it specific to MacOS-14 (i.e. on MacOS15 it works fine)

**Comment by Isalia20** [comment]
I'm on 14.6.1 macos and this outputs the same numbers(2 in this case) before the changes:
```
>>> probs = torch.tensor([[.25,.1],[.25,.1],[.25,.1],[.25,.7]]).T.to("mps")
>>> samples = {torch.multinomial(probs, 1).flatten()[0].item() for _ in range(200)}
>>> samples
{2}
```
Untitled

Regarding the workflow errors, I fixed the linter error sorry for that, others seem to be irrelevant for this PR

**Comment by Isalia20** [comment]
Same for 14.4.1 MacOS Sonoma

**Comment by malfet** [comment]
@Isalia20 makes sense, but do you mind restricting .contiguous() call to just MacOS-14?

File: aten/src/ATen/native/mps/operations/Distributions.mm [hunk 1]
```diff
@@ -636,7 +636,9 @@ Tensor normal_mps(const Tensor& mean, const Tensor& std, std::optional<Generator
     // s = argmax( p / (-log(eps)) ) where eps ~ U(0, 1).
     // We can also simplify the formula above by
     // s = argmax( p / q ) where q ~ Exp(1)
-    Tensor q = at::empty_like(self).exponential_(1, gen);
+    // Create `q` as contiguous tensor to ensure memory layout supports inplace operations
+    Tensor q = at::empty(self.sizes(), self.options().memory_format(c10::MemoryFormat::Contiguous));
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Distributions.mm:640]:
```suggestion
    const auto has_strided_api = is_macos_13_or_newer(MacOSVersion::MACOS_VER_15_0_PLUS);
    auto q = has_strided_api ? at::empty_like(self) : at::empty(self.sizes(), self.options().memory_format(c10::MemoryFormat::Contiguous));
```

File: aten/src/ATen/native/mps/operations/Distributions.mm [hunk 2]
```diff
@@ -636,7 +636,9 @@ Tensor normal_mps(const Tensor& mean, const Tensor& std, std::optional<Generator
     // s = argmax( p / (-log(eps)) ) where eps ~ U(0, 1).
     // We can also simplify the formula above by
     // s = argmax( p / q ) where q ~ Exp(1)
-    Tensor q = at::empty_like(self).exponential_(1, gen);
+    // Create `q` as contiguous tensor to ensure memory layout supports inplace operations
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Distributions.mm:639]:
```suggestion
    // If needed, create `q` as contiguous tensor to ensure memory layout supports inplace operations
```

**Comment by Isalia20** [comment]
Oh, didn't know macos 15 supported it. Thanks for the suggestion

File: aten/src/ATen/native/mps/operations/Distributions.mm [hunk 3]
```diff
@@ -636,7 +636,10 @@ Tensor normal_mps(const Tensor& mean, const Tensor& std, std::optional<Generator
     // s = argmax( p / (-log(eps)) ) where eps ~ U(0, 1).
     // We can also simplify the formula above by
     // s = argmax( p / q ) where q ~ Exp(1)
-    Tensor q = at::empty_like(self).exponential_(1, gen);
+    // If needed, create `q` as contiguous tensor to ensure memory layout supports inplace operations
+    const auto has_strided_api = is_macos_13_or_newer(MacOSVersion::MACOS_VER_15_0_PLUS);
+    auto q = at::empty_like(self, has_strided_api ? MemoryFormat::Preserve : MemoryFormat::Contiguous);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Distributions.mm:641]:
```suggestion
    auto q = at::empty_like(self, {}, has_strided_api ? std::nullopt : std::optional(MemoryFormat::Contiguous));
```

**Comment by Isalia20** [comment]
I guess we can merge, or do we need to rerun the workflow?

**Comment by Isalia20** [comment]
anything new on this?

**Comment by Isalia20** [comment]
Any updates @malfet ?

**Review by malfet — APPROVED** [review]
